geoutils/raster/delayed.py

```diff
@@ -0,0 +1,777 @@
+        return rio.coords.BoundingBox(*rio.transform.array_bounds(self.height, self.width, self.transform))
+
+    def bounds_projected(self, crs: rio.crs.CRS = None) -> rio.coords.BoundingBox:
+        if crs is None:
+            crs = self.crs
+        return _get_bounds_projected(bounds=self.bounds, in_crs=self.crs, out_crs=crs)
+
+    @property
+    def footprint(self) -> gpd.GeoDataFrame:
+        return _get_footprint_projected(self.bounds, in_crs=self.crs, out_crs=self.crs, densify_points=100)
+
+    def footprint_projected(self, crs: rio.crs.CRS = None) -> gpd.GeoDataFrame:
+        if crs is None:
+            crs = self.crs
+        return _get_footprint_projected(self.bounds, in_crs=self.crs, out_crs=crs, densify_points=100)
+
+    @classmethod
+    def from_dict(cls: type[GeoGridType], dict_meta: dict[str, Any]) -> GeoGridType:
+        """Create a GeoGrid from a dictionary containing transform, shape and CRS."""
+        return cls(**dict_meta)
+
+    def shift(
+        self: GeoGridType,
+        xoff: float,
+        yoff: float,
+        distance_unit: Literal["georeferenced"] | Literal["pixel"] = "pixel",
+    ) -> GeoGridType:
+        """Shift into a new geogrid (not inplace)."""
+
+        if distance_unit not in ["georeferenced", "pixel"]:
+            raise ValueError("Argument 'distance_unit' should be either 'pixel' or 'georeferenced'.")
+
+        # Get transform
+        dx, b, xmin, d, dy, ymax = list(self.transform)[:6]
+
+        # Convert pixel offsets to georeferenced units
+        if distance_unit == "pixel":
+            # Multiplying the offset by the resolution might amplify floating precision issues
+            # xoff *= self.res[0]
+            # yoff *= self.res[1]
+
+            # Using the boundaries instead!
+            xoff = xoff / self.shape[1] * (self.bounds.right - self.bounds.left)
+            yoff = yoff / self.shape[0] * (self.bounds.top - self.bounds.bottom)
+
+        shifted_transform = rio.transform.Affine(dx, b, xmin + xoff, d, dy, ymax + yoff)
+
+        return self.from_dict({"transform": shifted_transform, "crs": self.crs, "shape": self.shape})
+
+
+def _get_block_ids_per_chunk(chunks: tuple[tuple[int, ...], tuple[int, ...]]) -> list[dict[str, int]]:
+    """Get location of chunks based on array shape and list of chunk sizes."""
+
+    # Get number of chunks
+    num_chunks = (len(chunks[0]), len(chunks[1]))
+
+    # Get robust list of chunk locations (using what is done in block_id of dask.array.map_blocks)
+    # https://github.com/dask/dask/blob/24493f58660cb933855ba7629848881a6e2458c1/dask/array/core.py#L908
+    from dask.utils import cached_cumsum
+
+    starts = [cached_cumsum(c, initial_zero=True) for c in chunks]
+    nb_blocks = num_chunks[0] * num_chunks[1]
+    ixi, iyi = np.unravel_index(np.arange(nb_blocks), shape=(num_chunks[0], num_chunks[1]))
+    block_ids = [
+        {
+            "num_block": i,
+            "ys": starts[0][ixi[i]],
```

Comment:
took me a bit of time to understand the meaning of "s" and "e". Would it not be more straightforward to call them "ystart" and "yend" (or "ystop")?

Reply:
Added a comment to explain, trying to minimize the side of these nested dictionaries that can blow up memory fast for big dask graphs

Reply:
It's also described again right below.